### ISSUE TITLE
help strings

### ISSUE THREAD

**Issue by echatav** [Gabriel439/Haskell-Optparse-Generic-Library#5]
It might be nice to support help messages. I saw on the Reddit thread people mentioning using fields tagged by type level strings. I played with that.

``` haskell
newtype (<?>) (field :: *) (help :: Symbol) = Helped { getField :: field } deriving Generic
instance Show field => Show (field <?> help) where show = show . getField

parseHelpedField :: forall field help. (Typeable field, Read field, KnownSymbol help) => Maybe Text -> Parser (field <?> help)
parseHelpedField m = do
    let metavar = map toUpper (show (Data.Typeable.typeRep (Proxy :: Proxy field)))
    case m of
        Nothing   -> do
            let fs =  Options.metavar metavar
                    <> Options.help (symbolVal (Proxy :: Proxy help))
            Helped <$> Options.argument auto fs
        Just name -> do
            let fs =  Options.metavar metavar
                    <> Options.long (Data.Text.unpack name)
                    <> Options.help (symbolVal (Proxy :: Proxy help))
            Helped <$> Options.option   auto fs

instance KnownSymbol help => ParseField (Double <?> help) where parseField = parseHelpedField
instance KnownSymbol help => ParseFields (Double <?> help)
instance KnownSymbol help => ParseRecord (Double <?> help)
```

Then an example

``` haskell
data Command = Command
  { double :: Double <?> "it's a double!"
  , str :: String
  } deriving (Generic,Show)
instance ParseRecord Command
main = do
  x <- getRecord "test"

  print (x :: Command)
```

looks like

``` bash
Usage: test.hs --double DOUBLE --str STRING

Available options:
  -h,--help                Show this help text
  --double DOUBLE          it's a double!
```

I can see why this is not ideal because you have to use `getField` later but that seems better than having no help messages at all.

**Comment by Gabriella439** [comment]
Yeah, I'm fine with adding this, but this needs to be done in a more modular way so that we don't have write a separate instance for each type that we wish to add help information to

**Comment by echatav** [comment]
Since `field` and `field <?> help` are different types, they necessarily have to have separate instances. You could probably cut down on some of the boiler plate by factoring out some of the code shared by all the `parseField` instantiations and the `parseHelpedField` function above. Maybe there's a clever way to define instances like `instance (ParseField field , KnownSymbol help) => ParseField (field <?> help)` in one swoop. Anyway, Library users wouldn't have to define separate instances, just `Generic` as in the example above.

**Comment by echatav** [comment]
Closing, fixed by #6
